tinygrad/codegen/linearizer.py

```diff
@@ -428,29 +436,28 @@ def linearize(self):
     return self
 
   def ast_parse(self, x:LazyOp, acc: List[UOp], offs:Optional[List[int]], loaded_buffers:Dict[Union[MemBuffer, ConstBuffer, LocalBuffer], List[UOp]],\
-                do_reduce=False, loop_ctx=tuple(), cache=None, insert_before:Optional[UOp]=None) -> List[UOp]:
+                reduced_ops: Dict[LazyOp, List[UOp]]={}, do_reduce=False, loop_ctx=tuple(), cache=None) -> List[UOp]:
     if cache is None: cache = {}
     if x in cache: return cache[x]
     if x.op in BufferOps: return loaded_buffers[x.arg]
-    if x.op is UnaryOps.CAST: return [self.uops.add(UOps.BITCAST if x.arg[1] else UOps.CAST, self.get_base_dtype(x.arg[0]), (u,), x.arg[0], \
-                  insert_before=insert_before) for u in self.ast_parse(x.src[0], acc, offs, loaded_buffers, insert_before=insert_before)]
+    if x.op is UnaryOps.CAST: return [self.uops.add(UOps.BITCAST if x.arg[1] else UOps.CAST, self.get_base_dtype(x.arg[0]), (u,), x.arg[0]) \
+                                       for u in self.ast_parse(x.src[0], acc, offs, loaded_buffers, reduced_ops)]
     if x.op in ReduceOps and not do_reduce:
```

Comment:
don't we want to do this read also if do_reduce and x is not the reduce we want to do here?

Reply:
I'm not sure exactly what you mean but it seemed like the only time `do_reduce` was set to true was when `ast_parse` was directly fed the reduceop:

[linerizer.py:256](https://github.com/tinygrad/tinygrad/blob/267bbb57f9e767cd39804acbf0e0334031c12d35/tinygrad/codegen/linearizer.py#L256) `self.ast_parse(reduceop, acc, self.acc_offsets(self.full_buf_index), loaded_buffers, do_reduce=True, loop_ctx=loop_ctx)`

[linearizer.py:298](https://github.com/tinygrad/tinygrad/blob/267bbb57f9e767cd39804acbf0e0334031c12d35/tinygrad/codegen/linearizer.py#L298) `self.ast_parse(LazyOp(reduceop.op, (LazyOp(BufferOps.LOAD, (), self.bufs[-1]),)), acc, self.acc_offsets(-1), loaded_buffers, do_reduce=True, loop_ctx=loop_ctx)  # noqa: E501`

I agree, we do want to do that read, but in that case we'd have to pass the reduceop we want to do to `ast_parse` as well
I suppose that's not that hard, we just add another parameter or we could make `do_reduce` an `Optional[LazyOp]` and just pass it there

Reply:
> it seemed like the only time do_reduce was set to true was when ast_parse was directly fed the reduceop

on the second-pass reduce, ast_parse will recurse down to the first-pass reduce's reduceop, right?

Reply:
yeah but `do_reduce` will get turned off when it recurses through

`values = [self.ast_parse(v, acc, offs, loaded_buffers, reduced_ops, loop_ctx=loop_ctx, cache=cache) for v in x.src]`

also re: #4419 : `reduced_ops` is my way of accomplishing what I think you do with `accs`, I think replacing `accs` is prettier esp if it helps with parallel reduceops